### DISCUSSION TITLE
Welcome to dewheater Discussions!

### DISCUSSION THREAD

**Discussion by Curtis2012** [Curtis2012/dewheater#1] [Announcements]
Discussions relating to Dew Heater. For bugs, please use Issues instead.

**Comment by ON2ON-HAM** [comment]
Hi,
Installation went fine, but only 2 error ,, if you would have a minute to check out my issue, it would be great : 

![afbeelding](https://user-images.githubusercontent.com/27927522/151657153-97fd014c-9d6b-442e-9aeb-ffb0515afe80.png)


Thanks,
73 ON2ON Luke

**Reply by Pedritoprint** [reply]
The first Error is a GPIO ERROR i think and could be ignored. Do you have an other GPIO-channel in use like i have? Servo-control, Weather Sensor and Heater-Relais....
73 de HB9CMI

**Reply by Curtis2012** [reply]
@ON2ON-HAM

The first error is very common w Rpi's if a process has exited without releasing GPIO resources (for example: it crashed). It can usually be ignored unless you have other processes using GPIO and there is in fact a conflict.

Installing the GPIO library should resolve your second error. 

https://www.raspberrypi-spy.co.uk/2012/05/install-rpi-gpio-python-library/

**Comment by Curtis2012** [comment]
@ON2ON-HAM

I think we addressed these issues over on the dewheater discussion under the allsky repository, right?

**Comment by clint70** [comment]
This is a great addition for allsky, I need a little more details, if possible?  The relay that is being used?  Does it use 5v?  If so, does the 5v come from the RPi?  OR is it 12v from a sepreate power supply??  What relay did you use?  AND is there a basic block diagram with pinouts etc?

**Reply by Curtis2012** [reply]
Thanks Clint, the relay I used is controlled by 3-3.3V input (the Rpi GPIO pins are 3.3V). It switches 5V from the Rpi power pin.

**Reply by Curtis2012** [reply]
Im away from good internet service, but will try and post the relay specs. The relay I used swiches on 3-3.3V and I switch 5V power to the dew heater. It could drive 12V also, but you would need a 12V power source.

**Reply by Curtis2012** [reply]
I have not drawn up a diagram, but it is simple. I switch the relay using a  GPIO pin (see config file for which pin I used). The power input to be switched is from the Rpi 5V pin.

**Reply by Curtis2012** [reply]
Relay used:

https://www.amazon.com/gp/aw/d/B08W3XDNGK?psc=1&ref=ppx_pop_mob_b_asin_title

**Reply by Curtis2012** [reply]
You should also be able to drive a resistor based heater circuit too. I think that is how Pedritoprint (who posted above) is using it.

**Comment by Shand359** [comment]
Can I please ask, how do I install this and set it up. I have no background in Linux or RPi programming. I have a round dewheater I’d like to fit if possible 

I have also just found out that the ring runs from 12V supply

**Reply by Shand359** [reply]
I’m guessing I might need to try sourcing a separate  5V dewheater….

**Reply by Curtis2012** [reply]
Shand, as built my design drives a 5V dew heater (see the description in the repository). I used a USB powered dew heater I bought on Amazon.

You could drive a 12V dew heater, but would need to add a 12V power source to your build.

**Reply by Curtis2012** [reply]
Re install. I wrote up an install procedure which I have not posted yet. I am away from good Internet service and not sure that I can access the write up from here, but I will try.


That said, the set up is pretty simple. There are few dependencies. Basically just build it, create the service definition, and start the service. 

I will see if I can upload my write up. If not, I will post some simplified instrictions.

**Reply by ON2ON-HAM** [reply]
@Shand359 maybe this could be already helpfull,, the relay type needed, and as you can see in the dewheaterconfig.json file, the pin(s) are set where to connect your DHT11/22 and where to connect the Relay. The relay is very simple ,,, one pin to the GPIO pin 23, one to the GPIO Ground, and one to the GPIO 5volt. The DHT11/22 sensor, goes to GPIO pin 4 , and also to ground and 5 volt.  Basically that is it. The relay will switch whatever you connect to it, depending on the settings in dewheaterconfig.json regarding dew, and temperature max/min. 
![afbeelding](https://user-images.githubusercontent.com/27927522/163788623-37561b2b-c40e-4391-ba05-869cad112512.png)

**Reply by Curtis2012** [reply]
There ya go! Its as simple as that. 

Thanks for the write up @ON2ON-HAM !

**Reply by clint70** [reply]
Can't say thank you enough... this thread is great!  Appreciate the time from both of you Curtis2012 & ON2ON-HAM.  Now to get all this put together ;)  I have the case, RPi loaded with Allsky and it has been running without the dew heater (moisture on the dome driving me nuts).  :D  Clear views ahead...

Cheers!

**Reply by Curtis2012** [reply]
@clint70 

De nada! Glad to be of help and to see others making good use of this module. 

Please update us on how it goes for you.

**Reply by Curtis2012** [reply]
One more note I thought of: for most configs you use the NO port on the relay. Most relays are controlled by receiving a +V signal. Some are controlled by a -V signal. The invertOnOff parameter = true is used for -V triggered relays. Ive not tested this personally, at least not with an actual -V triggered relay, but I think that @Pedritoprint is running a -V triggered relay.

**Reply by clint70** [reply]
Thank you for the insight.  Once I get the relay and install, I will keep that info close.  Thought I had a relay, but had to order one, once here and start the install, I will update.  Cheers!

**Comment by ON2ON-HAM** [comment]
Don't forget to paint the LED of the relay board black ;)
The red led light gets into your image, if any tiny holes are between the
lens opening and the bottom of your case were the relay resides.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Op wo 20 apr. 2022 om 11:38 schreef clint70 ***@***.***>:

**Comment by clint70** [comment]
Hi Curtis, just finally getting around to installing the dew heater, question about the DHT11 vs a BME280, should work?  I'm gonna see if It will be able to do the same.  NO DHT11 rn

:D  Thanks for the great project.

**Reply by clint70** [reply]
update... bme280 running and service is good.  BTW love your code, easy to follow, great structure.  THANK YOU!

**Reply by Curtis2012** [reply]
Thanks Clint!

I was a pro developer back in the day...

**Reply by Curtis2012** [reply]
Glad u are up and running.

I think you and @Pedritoprint are both running the BME280.

**Reply by clint70** [reply]
I would have used a DHT11 if I had one around, but I have a ton of the BMEs, so easier.  AND TBH, was not a big deal to update your code for the BME.

Now I would like to see if I can get the output (Temp, Humid, Press) on the website ;)  1 step at a time ... glad I got it running today.
Cheers!

**Reply by Pedritoprint** [reply]
> Glad u are up and running.
> 
> I think you and @Pedritoprint are both running the BME280.

That is correct, i have also the BME280 in use.

**Reply by Curtis2012** [reply]
@clint70 I would like to add BME380 support to the code. Would you start a branch to add it?

**Reply by clint70** [reply]
For sure, I'll set it up tonight.

**Reply by Curtis2012** [reply]
Cool. Thanks.

**Reply by clint70** [reply]
@Curtis2012, tried to publish branch, but I don't have permission.  If you want I can create a fork?  Let me know.

**Reply by Curtis2012** [reply]
Clint, I will change permissions so that you can create a branch.

**Reply by Curtis2012** [reply]
@clint70 Invite sent!

**Comment by Curtis2012** [comment]
@clint70 

Thanks for creating the BME280 branch. I would like to create a version based upon that which supports either the DHT or the BME as a configuration option. Are you OK with that?

**Reply by clint70** [reply]
Absolutely, sorry I didn't do that.  Would make sense...  Also I had some troubles with python versions, mostly me and configuring the Pi.  BUT the version I have working right now is Python 2.  NOT a big issue.  

For some reason the print statement with 'end' in it would not work either.  I am curious why, and if there is something I can do, let me know too.  I'm not particularly good a programming, but I can 'hack' mostly.  Thanks for letting me participate.

**Reply by Curtis2012** [reply]
Clint, I will try and take  a look this week and start make this modification. 

Re Ptyhon, unless you have a compelling reason to run 2, I would upgrade to the most current version. Dew Heater was built under 3.7.  So some issues you encounter could be due to Python version.

**Reply by clint70** [reply]
Ya... I know that python2 is probably not a great plan, I was just going for 'easiest' way to get going.  I am on holidays with family and the allsky is down for maint.  Once I get home, I'll fix the allsky and clean up the dewheater code.  

Looking at trying to add a second sensor for outside the case for dew point calculation... but that is just a curiosity right now.  Might look into it more, see if it is worth the effort

**Reply by Curtis2012** [reply]
No worries, its all hobby stuff. 

I have considered adding a second sensor. This would allow the dew point calculation to be made, or modified, based upon external ambient conditions in additional to internal (under the dome).

**Reply by Curtis2012** [reply]
@clint70 

I created a branch of your BME branch named "merge_BME208_and_DHT_support". I will integrate parts of both code sets here.

I've ordered a BME280 so that I cant test locally.

**Reply by clint70** [reply]
Ok... sounds great!  I'll look at what you have done.  Let me know if there is anything I can do.

**Reply by Curtis2012** [reply]
I've just started restructuring the code to accommodate both sensor types. Its a relatively minor change really. 

It will be at least 10 days before I get a BME sensor here in Panama. At that point I will do some set up and testing. It would be good if you could also do some testing after I do the initial testing.

**Reply by clint70** [reply]
No problem at all!  Just let me know.  And I'll do any testing necessary. 

Ya, I imagine to integrate both would not be too difficult for you ;).  I'll go look at the code. 

Thanks for including me 😄

**Reply by Curtis2012** [reply]
@clint70 Well....civil unrest here in Panama has disrupted a lot of things including imports. So my BME280 is delayed.

**Reply by clint70** [reply]
I hope you are ok... and things are not too worrisome.   I am not on the computer as much atm, and the Allsky is down for a redesign of the case, the one I am using heats up inside.  The RPi hits 83 degrees :O ... 

I think I know what you are trying to do with for the code, might play around over the next few weeks or so.  Take care, hope you are well.

**Reply by Curtis2012** [reply]
All good here, just inconvenient (roads are blocked so fuel trucks cant get to gas stations, etc).

Yep,, heat is problem in full sun. Ive been working on an automatic dome cover.

**Reply by Curtis2012** [reply]
@clint70 

Re what I want to do with the code. I want the sensor type to be configurable. A field in the config file indicates the sensor type and then the code uses the appropriate sensor. I've made most of those changes already. I will make sure the branch is updated with those changes so you can see them.

**Reply by Curtis2012** [reply]
@clint70 

I've pushed my changes up. The only 2 files I've changed are dewheater.py and dewheaterconfig.json. These changes are incomplete awaiting a BME280 I can test with. 

There are several related utilities that will need to be updated too. These are dewheateron.py, dewheateroff.py, and dewheatertest.py. If you wanted to help out by starting changes on those that would be great.

**Reply by clint70** [reply]
For sure, I'll look this week.  I have the BME so I can test etc too.  IF I get stuck, I'll reach out for input.  Take it easy.

**Reply by Curtis2012** [reply]
Good. Thanks!

**Reply by Curtis2012** [reply]
@clint70 

Protests settled down and I finally got my BME board. I've connected it to my allskycam Rpi and have started on completing the mods to dewheater.py. I will leave both the DHT and BME connected so I've got a complete test platform.

The only utility that needs to be modified is dewheatertest.py. Let me finish up the initial code on dewheater.py and then you can use that to update dewheatertest.py. Should only be a few days for me to complete.

**Reply by clint70** [reply]
Wow! I'm not moving quickly on this, sorry.  Been over run by family events.  I have my Allsky down, but it still works, so I can do testing... 

This is awesome, thx for doing this. 

Cheers

**Reply by Curtis2012** [reply]
No rush, its just hobby stuff.

I've restructured the code a little. Created DHT and BME objects which makes it cleaner. Doing initial debugging now.

**Reply by clint70** [reply]
awesome

**Reply by Curtis2012** [reply]
My Allsky Cam is up and running with BME/DHT version of dewheater. I have both a BME and a DHT installed so I can switch for testing just using the config file. Appears to be working fine. Will test for a few days, make some code tweaks, and then the final commit and merge up to main. I have pushed the most recent code up to the "merge_BME208_and_DHT_support" branch. 

![image](https://user-images.githubusercontent.com/41484582/182156979-eee7fe77-2ee4-4c81-af2d-c592886cb436.png)

**Reply by Curtis2012** [reply]
I pushed the latest version of the DHT/BME code to the "merge_BME280_and_DHT_support" branch. It has been up and running on my allskycam for about a week now, but could use more testing (especially in a colder environment).

Open tasks:

- update Readme.MD
- update dewheatertest.py to new dual sensor code
- update wiring diagrams to show for both sensors
- accuracy testing. Based on local conditions from my weather station the sensors are in the ball park, but I have not tested in detail yet. 

Volunteers welcome!

**Reply by Dentonknifeworks** [reply]
Well getting this error: Tried redoing it a few times and same issue.
My other code is working with reading the BME280 so I know it's wired correctly

Traceback (most recent call last):
  File "/home/pi/dewheater/dewheater.py", line 112, in <module>
    bme = BME820Class()
  File "/home/pi/dewheater/dewheater.py", line 103, in __init__
    self.bme_calibration_params = bme280.load_calibration_params(self.bus, self.address)
AttributeError: module 'bme280' has no attribute 'load_calibration_params'

**Reply by Curtis2012** [reply]
Will take a look. Its up and running with the BME on my allskycam so not sure why you would get that. Looks like an incorrect version of the code, maybe something is out of sync on GitHub.

**Reply by Dentonknifeworks** [reply]
I also found it odd that I had to change the case of Adafruit_DHT to adafruit_dht to get it to compile.
I started that out as a new SD card and install and Still have to find a good source for the Adafruit_DHT I  had that issue before and noted the page that worked last time. I am more the willing to start from scratch for like the 10th time LOL

**Reply by Curtis2012** [reply]
@Dentonknifeworks 

Looks like a BME280 library difference. This code was built using the library below:

https://github.com/rm-hull/bme280

You can check which BME280 library you are running using the command below:

pi@allsky:~ $ pip list | grep bme280

If you have the above library you should see the output below:

RPi.bme280          0.2.4

If you don't then install it and that should resolve the issue. It should not conflict with other BME libraries.

**Reply by Curtis2012** [reply]
> @Dentonknifeworks
> 
> Looks like a BME280 library difference. This code was built using the library below:
> 
> https://github.com/rm-hull/bme280
> 
> You can check which BME280 library you are running using the command below:
> 
> pi@allsky:~ $ pip list | grep bme280
> 
> If you have the above library you should see the output below:
> 
> RPi.bme280 0.2.4
> 
> If you don't then install it and that should resolve the issue. It should not conflict with other BME libraries.

Will also take a look that the DHT library issue...

**Reply by Dentonknifeworks** [reply]
pi@DewHeater:~ $ pip list | grep bme280
adafruit-circuitpython-bme280    2.6.13
bme280                           0.6
RPi.bme280                       0.2.4
pi@DewHeater:~ $ 

Look correct on my end

**Reply by Curtis2012** [reply]
That's  the Ada Fruit version of the library. Which is fine, but not the one used in this case. If you install the library below then it should resolve the issue. 

https://github.com/rm-hull/bme280

**Reply by Dentonknifeworks** [reply]
Ok got it working a new SD card
install pip3 install adafruit-circuitpython-dht
Change Adafruit_DHT in code to adafruit_dht  2 places
install RPi.bme280: pip3 install RPi.bme280
install bme280 pip3 install bme280
install smbus2: pip3 install smbus2
enable I2C interface: sudo raspi-config

Now working


sudo pip3 install Adafruit_DHT  errors out on me every time.

**Reply by Dentonknifeworks** [reply]
Well, I restated the pi this morning now getting.  if conditions.dewPointMet:
AttributeError: 'ConditionsClass' object has no attribute 'dewPointMet'

Wow I have no luck at all LOL sorry I just don't understand why this should be going this way seems that others are not having issues.

**Reply by Curtis2012** [reply]
No problem, this is good...finding issues is the purpose of testing! In your case they seem to be environmental issues, but others will run into them at some point too so the questions will already be answered. 

I will take a look soon at the "ConditionsClass" issue.

**Reply by Dentonknifeworks** [reply]
"dewHeaterCutoutOffset": 15.0,  That was what was causing the ConditionsClass issue.
I have that on the DHT22 and it works I had copied that to the BME280 thinking I would be using the same offsets.
Back to working now 

My new workflow from new install of Bullseye 
enable I2C interface: sudo raspi-config
pip3 install adafruit-circuitpython-dht
pip3 install RPi.bme280
pip3 install smbus2
pip3 install meteocalc
Change Adafruit_DHT in code to adafruit_dht  2 places

Brad

**Reply by Curtis2012** [reply]
Interesting. Glad its working for you now. I will take a look at the CutoutOffset between the two sensors.

**Reply by Curtis2012** [reply]
In reviewing the code I think there are some configuration settings scenarios where the "dewPointMet" variable might not get defined on start up (one down side of interpreted languages vs compiled is that variables are defined dynamically). I will add a line of code to force it to be defined. 

The CutIn/CutOut values should not differ by sensor. These are intended to allow cutin/cutout points to be offset from the actual dew point to compensate for installation differences. Will review.

**Reply by Curtis2012** [reply]
@Dentonknifeworks 

I added code to explicitly define "dewPointMet" during initialization of the Conditions class. This change was made in the "merge_BME280_and_DHT_support"  branch. Clone this branch and update your install (the actual change was only to the dewheater.py file so you could just update that one).

Congratulations! This was actually a condition that was not covered by the code, and never has been. You helped squash a well hidden bug!

**Reply by Curtis2012** [reply]
@clint70 

FYI: I accidentally pushed the code from the "merge_BME280_and_DHT_support" branch to the "BME280_support" branch. Since the former supersedes the later anyway its not a big deal, but just wanted you to know.

**Reply by Dentonknifeworks** [reply]
Tested and your change worked.
I added an extra text function for the extra text in Allsky.

def fileouput():   
    file = open('sensor_readings.txt', 'w')
    file.write('Inside Dome\n')
    file.write("Temp = %3.1fC\nHumidity = %3.1f\nDew Point = %3.1fC\nHeater = %s " % (
        conditions.temperature, conditions.humidity, conditions.dewPoint.c, conditions.dewPointMet))
    file.close()

**Reply by Curtis2012** [reply]
Good. I will put that function to use when I follow up on adding logging. 

I also plan to add support for dual sensors of either type. Example: one under the dome and one in the housing or outside. Which is used to control the dew heater will be configurable.

**Reply by Dentonknifeworks** [reply]
I am using a Sqllite DB and code that is storing the BME280 data. Looking into what to use for charting. I used Node-Red on the DHT it worked but seemed overkill for the need so now looking for something less complicated.
